torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -531,44 +528,41 @@ class CppMicroGemmAMX(CppMicroGemm):
     alignas(4096) {{input_t}} dequantized_B_buf[buf_size];
     {%- endif %}
 
-    const auto b_tile_ptr_stride = ldb * {{vnni_size}};
-
-    auto load_B_row = [&]({{input2_t}}* {{restrict_keyword}} src, {{input_t}}* {{restrict_keyword}} dst) {
-        auto b_int8 = at::vec::Vectorized<int8_t>::loadu(src, static_cast<int64_t>(32));
-        auto b_bf16 = at::vec::convert<{{input_t}}>(b_int8);
-        b_bf16.store(dst);
-    };
-
-    auto load_B_tile = [&]({{input2_t}}* B_ptr, int idx, int num_b_rows) {
-        {{input_t}}* base_addr = dequantized_B_buf + idx;
-        {{kernel.unroll_pragma(8)}}
-        for (int i = 0; i < num_b_rows; i++) {
-            load_B_row(
-                B_ptr + i * b_tile_ptr_stride,
-                base_addr + i * 32
-            );
-        }
-    };
-    auto load_dequantized_B = [&](int n) {
+    auto load_dequantized_B = [&](int base_idx) {
         // Load a tile of B & cache it in L1D.
-        {{kernel.unroll_pragma(4)}}
-        for (int k = 0; k < K; k += {{block_k}}) {
-            int num_b_rows = (k < last_k_offset) ? 16 : tail_k_size;
-            {{kernel.unroll_pragma(2)}}
-            for (int tile_col = 0; tile_col <= 1; tile_col++) {
-                load_B_tile(
-                    const_cast<{{input2_t}}*>(B) + n + k * ldb + tile_col * {{16 * vnni_size}},
-                    (k / {{block_k // 2}} + tile_col) * num_elements_per_b_tile,
-                    num_b_rows
-                );
-            }
+        // The assumption here is that if N would be a multiple of block_n.
+        {{input2_t}}* base_addr = const_cast<{{input2_t}}*>(B) + base_idx;
+        for (int idx_dq = 0, idx_q = 0; idx_dq < buf_size; idx_q += ldb, idx_dq += {{block_n}}) {
```

Comment:
Thanks for the PR. I am wondering what's the case when `ldb` != `block_n`? It seems we only handle the first number of `block_n` elements.

Reply:
Hi @leslie-fang-intel, as discussed offline, this code assumes weight-blocking as in https://github.com/pytorch/pytorch/pull/140258#discussion_r1841341702.

The code before today's commits assumed https://github.com/pytorch/pytorch/pull/140258#discussion_r1836285920.

Reply:
Thanks for the explanation. I am not sure about the scenario that `ldb` != `block_n`. To understand if this line of code ` for (int idx_dq = 0, idx_q = 0; idx_dq < buf_size; idx_q += ldb, idx_dq += {{block_n}}) {`, I guess we need to know the scenario. What do you think?

Reply:
> It seems we only handle the first number of `block_n` elements

@leslie-fang-intel, can you please elaborate on this statement? Thanks!

Reply:
> > It seems we only handle the first number of `block_n` elements
> 
> @leslie-fang-intel, can you please elaborate on this statement? Thanks!

Maybe I didn't understand the scenario for why `idx_q` step by `ldb` but `idx_dq` step by `block_n`. Can you describe more about the scenario why the step of `idx_q` and `idx_dq` is different?

Reply:
> I guess we need to know the scenario. What do you think?

As [discussed here](https://github.com/pytorch/pytorch/pull/140258#discussion_r1841308006), I think this case isn't currently possible due to the current weight-packing implementation.

> Can you describe more about the scenario why the step of `idx_q` and `idx_dq` is different?

This scenario requires different weight-blocking implementation, as in https://github.com/pytorch/pytorch/pull/140258#discussion_r1841341702

The `ldb` in that figure is `N` but `N != block_n`.

Reply:
> since the weight has been packed into `[padded_n // block_n, k, block_n]`

Do you know why [this scenario](https://github.com/pytorch/pytorch/pull/140258#discussion_r1836285920) is not true? Thanks!

<img width="598" alt="image" src="https://github.com/user-attachments/assets/12270c7c-256d-479d-8b88-66ab45ab4a88">

Reply:
- We currently pack weight into `[padded_n // block_n, k, block_n]`
- We get the `ldb` of micro gemm input as `ldb = kernel.stride(B, 0)` https://github.com/pytorch/pytorch/blob/b379a28a95aa0b0d167f08b8d70c8b85a248309d/torch/_inductor/codegen/cpp_micro_gemm.py#L142

So, I think per current implementation, `block_n` should equal to `ldb`. Nevertheless, as @jgong5 mentions here for API semantics: https://github.com/pytorch/pytorch/pull/140258#discussion_r1839529082

> The API semantics is that B is of the shape [K, N] with stride [ldb, 1] where ldb is not necessarily equal to N. I don't think it is correct to assume B is contiguous.

I guess using `ldb` to `idx_q += ldb` should be fine.

Reply:
> the micro-gemm makes any assumption that the weight is "blocked" at the upper level, right

@jgong5, yes, the weight-packing at upper level is transparent to the _current_ micro-kernel implementation, which uses `N == block_n`. 

But if `N != block_n`, then the weight-packing format seems to matter to the micro-kernel,

Reply:
Do we need to change this line if `use_cached_dequantized_B`?
https://github.com/pytorch/pytorch/blob/c79e78b5034198f9d6801b4fef710b9b9b0e9193/torch/_inductor/codegen/cpp_micro_gemm.py#L593

Reply:
That code pertains to calling the inner kernel in the micro-kernel.
So, no change is required, because during loading of AMX tiles in the inner-kernel, the dequantized B cache would have the same `ldb` as `B`.

Basically, with this implementation, the elements of `B` have the same indices in dequantized B cache.

Reply:
@jgong5, please help resolve a confusion - can the AMX micro-kernel be used even if the weight is not packed with `[N//block_n, K, block_n]` layout? I mean does it simply need VNNI layout without requiring weight-prepacking `[N//block_n, K, block_n]` one level above it? Thanks!